src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java

```diff
@@ -375,27 +375,26 @@ public void registerToolchains(Sequence<?> toolchainLabels) throws EvalException
       },
       useStarlarkThread = true)
   public ModuleExtensionProxy useExtension(
-      String extensionBzlFile, String extensionName, boolean devDependency, StarlarkThread thread)
-      throws EvalException {
+      String extensionBzlFile, String extensionName, boolean devDependency, StarlarkThread thread) {
     ModuleExtensionProxy newProxy =
         new ModuleExtensionProxy(extensionBzlFile, extensionName, thread.getCallerLocation());
 
     if (ignoreDevDeps && devDependency) {
       // This is a no-op proxy.
-      return newProxy;
+      return newProxy.withDevDependency(devDependency);
     }
 
     // Find an existing proxy object corresponding to this extension.
     for (ModuleExtensionProxy proxy : extensionProxies) {
       if (proxy.extensionBzlFile.equals(extensionBzlFile)
           && proxy.extensionName.equals(extensionName)) {
-        return proxy;
+        return proxy.withDevDependency(devDependency);
```

Comment:
I don't think this works. If you add a test like the following:

```python
proxy1 = use_extension(":foo.bzl", "proxy")
proxy1.tag()
proxy2 = use_extension(":foo.bzl", "proxy", dev_dependency=True)
proxy2.tag()
```

I believe the tag on `proxy2` will be lost. (This is because the Java object backing `proxy2` was never recorded anywhere; normally all proxies are recorded in `extensionProxies`)

What you need to do is make `devDependency` a parameter to the constructor of `ModuleExtensionProxy`, and in the `if` condition here, also check if `devDependency` matches. Then at the very end in `#buildModule`, you need to somehow merge two `ModuleExtensionUsage` objects if they only differ in `devDependency`. This is somewhat unwieldy but it's my first reaction.

Reply:
Your suggested approach is actually what I tried first, but merging the usages while preserving the orders of imports and tags proved to be rather difficult.

Instead, I am essentially creating proxy-proxies here (phew): Both `proxy.withDevDependency(true)` and `proxy.withDevDependency(false)` share the `imports` and `tags` of the original `proxy` and thus should make that test pass (see https://github.com/bazelbuild/bazel/pull/17909/files#diff-99af257732180398dadf7c29772bd194ca9518e59625e9ca10e3b1a31c5d9afaL553, albeit that one has a different order of dev/non-dev). Do you still see an issue here?

Reply:
oh, I see! I missed the fact that the proxy-proxies share the accumulated tags. Sorry for the noise :)

Reply:
alternatively we could formulate this a bit differently. Rename `ModuleExtensionProxy` to `ModuleExtensionUsageBuilder` or something and let it have a method `ModuleExtensionProxy getProxy(boolean devDependency)`. This way you cannot nest proxies in arbitrary depths.

Reply:
I agree that's the safer design, implemented it.